// @hathor/wallet-lib/lib/opcodes.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.OP_PUSHDATA1 = exports.OP_CHECKSIG = exports.OP_EQUALVERIFY = exports.OP_HASH160 = exports.OP_DUP = exports.OP_GREATERTHAN_TIMESTAMP = void 0;

var _buffer = require("./utils/buffer");

/**
 * Copyright (c) Hathor Labs and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

/**
 * Opcodes used to generate output script
 * @module Opcodes
 */

/**
 * Checks if timestamp is greater than a value
 */
const OP_GREATERTHAN_TIMESTAMP = (0, _buffer.hexToBuffer)('6f');
/**
 * Duplicates value
 */

exports.OP_GREATERTHAN_TIMESTAMP = OP_GREATERTHAN_TIMESTAMP;
const OP_DUP = (0, _buffer.hexToBuffer)('76');
/**
 * Calculates hash160 of value
 */

exports.OP_DUP = OP_DUP;
const OP_HASH160 = (0, _buffer.hexToBuffer)('a9');
/**
 * Verifies if values are equal
 */

exports.OP_HASH160 = OP_HASH160;
const OP_EQUALVERIFY = (0, _buffer.hexToBuffer)('88');
/**
 * Verifies signature
 */

exports.OP_EQUALVERIFY = OP_EQUALVERIFY;
const OP_CHECKSIG = (0, _buffer.hexToBuffer)('ac');
/**
 * Shows that pushdata will need length value
 */

exports.OP_CHECKSIG = OP_CHECKSIG;
const OP_PUSHDATA1 = (0, _buffer.hexToBuffer)('4c');
exports.OP_PUSHDATA1 = OP_PUSHDATA1;
